// astro/dist/core/build/scan-based-build.js
import { fileURLToPath } from "url";
import * as vite from "vite";
import { createBuildInternals } from "../../core/build/internal.js";
import { rollupPluginAstroBuildHTML } from "../../vite-plugin-build-html/index.js";
import { rollupPluginAstroBuildCSS } from "../../vite-plugin-build-css/index.js";
function entryIsType(type) {
  return function withPage([_, pageData]) {
    return pageData.route.type === type;
  };
}
function reduceEntries(acc, [key, value]) {
  acc[key] = value;
  return acc;
}
function routesOfType(type, allPages) {
  return Object.entries(allPages).filter(entryIsType(type)).reduce(reduceEntries, {});
}
async function build(opts) {
  const { allPages, astroConfig, logging, origin, pageNames, routeCache, viteConfig, viteServer } = opts;
  const internals = createBuildInternals();
  return await vite.build({
    logLevel: "error",
    mode: "production",
    build: {
      emptyOutDir: true,
      minify: "esbuild",
      outDir: fileURLToPath(astroConfig.dist),
      rollupOptions: {
        input: [],
        output: {
          format: "esm"
        }
      },
      target: "es2020"
    },
    plugins: [
      rollupPluginAstroBuildHTML({
        astroConfig,
        internals,
        logging,
        origin,
        allPages: routesOfType("page", allPages),
        pageNames,
        routeCache,
        viteServer
      }),
      rollupPluginAstroBuildCSS({
        internals
      }),
      ...viteConfig.plugins || []
    ],
    publicDir: viteConfig.publicDir,
    root: viteConfig.root,
    envPrefix: "PUBLIC_",
    server: viteConfig.server,
    base: astroConfig.buildOptions.site ? new URL(astroConfig.buildOptions.site).pathname : "/"
  });
}
export {
  build
};
