// astro/dist/vite-plugin-jsx/index.js
import babel from "@babel/core";
import esbuild from "esbuild";
import * as colors from "kleur/colors";
import * as eslexer from "es-module-lexer";
import path from "path";
import { error } from "../core/logger.js";
import { parseNpmName, resolveDependency } from "../core/util.js";
const JSX_RENDERER_CACHE = /* @__PURE__ */ new WeakMap();
const JSX_EXTENSIONS = /* @__PURE__ */ new Set([".jsx", ".tsx"]);
const IMPORT_STATEMENTS = {
  react: "import React from 'react'",
  preact: "import { h } from 'preact'",
  "solid-js": "import 'solid-js/web'"
};
const PREVENT_UNUSED_IMPORTS = ";;(React,Fragment,h);";
function getEsbuildLoader(fileExt) {
  return fileExt.substr(1);
}
async function importJSXRenderers(config) {
  const renderers = /* @__PURE__ */ new Map();
  await Promise.all(config.renderers.map((name) => {
    return import(resolveDependency(name, config)).then(({ default: renderer }) => {
      if (!renderer.jsxImportSource)
        return;
      renderers.set(renderer.jsxImportSource, renderer);
    });
  }));
  return renderers;
}
async function transformJSX({ code, mode, id, ssr, renderer }) {
  const { jsxTransformOptions } = renderer;
  const options = await jsxTransformOptions({ mode, ssr });
  const plugins = [...options.plugins || []];
  const result = await babel.transformAsync(code, {
    presets: options.presets,
    plugins,
    cwd: process.cwd(),
    filename: id,
    ast: false,
    compact: false,
    sourceMaps: true,
    configFile: false,
    babelrc: false
  });
  if (!result)
    return null;
  return {
    code: result.code || "",
    map: result.map
  };
}
function jsx({ config, logging }) {
  let viteConfig;
  return {
    name: "astro:jsx",
    enforce: "pre",
    configResolved(resolvedConfig) {
      viteConfig = resolvedConfig;
    },
    async transform(code, id, opts) {
      const ssr = Boolean(opts == null ? void 0 : opts.ssr);
      if (!JSX_EXTENSIONS.has(path.extname(id))) {
        return null;
      }
      const { mode } = viteConfig;
      let jsxRenderers = JSX_RENDERER_CACHE.get(config);
      if (!jsxRenderers) {
        jsxRenderers = /* @__PURE__ */ new Map();
        const possibleRenderers = await importJSXRenderers(config);
        if (possibleRenderers.size === 0) {
          throw new Error(`${colors.yellow(id)}
Unable to resolve a renderer that handles JSX transforms! Please include a \`renderer\` plugin which supports JSX in your \`astro.config.mjs\` file.`);
        }
        for (const [importSource2, renderer] of possibleRenderers) {
          jsxRenderers.set(importSource2, renderer);
        }
        JSX_RENDERER_CACHE.set(config, jsxRenderers);
      }
      if (jsxRenderers.size === 1) {
        const { code: jsxCode } = await esbuild.transform(code, {
          loader: getEsbuildLoader(path.extname(id)),
          jsx: "preserve",
          sourcefile: id,
          sourcemap: "inline"
        });
        return transformJSX({ code: jsxCode, id, renderer: [...jsxRenderers.values()][0], mode, ssr });
      }
      const { code: jsCode } = await esbuild.transform(code + PREVENT_UNUSED_IMPORTS, {
        loader: getEsbuildLoader(path.extname(id)),
        jsx: "transform",
        jsxFactory: "h",
        jsxFragment: "Fragment",
        sourcefile: id,
        sourcemap: "inline"
      });
      let imports = [];
      if (/import/.test(jsCode)) {
        let [i] = eslexer.parse(jsCode);
        imports = i;
      }
      let importSource;
      if (imports.length > 0) {
        for (let { n: spec } of imports) {
          const pkg = spec && parseNpmName(spec);
          if (!pkg)
            continue;
          if (jsxRenderers.has(pkg.name)) {
            importSource = pkg.name;
            break;
          }
        }
      }
      if (!importSource) {
        const multiline = code.match(/\/\*\*[\S\s]*\*\//gm) || [];
        for (const comment of multiline) {
          const [_, lib] = comment.match(/@jsxImportSource\s*(\S+)/) || [];
          if (lib) {
            importSource = lib;
            break;
          }
        }
      }
      if (importSource) {
        const jsxRenderer = jsxRenderers.get(importSource);
        if (!jsxRenderer) {
          error(logging, "renderer", `${colors.yellow(id)} No renderer installed for ${importSource}. Try adding \`@astrojs/renderer-${importSource}\` to your dependencies.`);
          return null;
        }
        const { code: jsxCode } = await esbuild.transform(code, {
          loader: getEsbuildLoader(path.extname(id)),
          jsx: "preserve",
          sourcefile: id,
          sourcemap: "inline"
        });
        return await transformJSX({ code: jsxCode, id, renderer: jsxRenderers.get(importSource), mode, ssr });
      }
      const defaultRenderer = [...jsxRenderers.keys()][0];
      error(logging, "renderer", `${colors.yellow(id)}
Unable to resolve a renderer that handles this file! With more than one renderer enabled, you should include an import or use a pragma comment.
Add ${colors.cyan(IMPORT_STATEMENTS[defaultRenderer] || `import '${defaultRenderer}';`)} or ${colors.cyan(`/* jsxImportSource: ${defaultRenderer} */`)} to this file.
`);
      return null;
    }
  };
}
export {
  jsx as default
};
