// astro/dist/core/preview/index.js
import http from "http";
import sirv from "sirv";
import { performance } from "perf_hooks";
import { fileURLToPath } from "url";
import * as msg from "../messages.js";
import { error, info } from "../logger.js";
import { subpathNotUsedTemplate, notFoundTemplate } from "../../template/4xx.js";
import { getResolvedHostForHttpServer } from "./util.js";
const HAS_FILE_EXTENSION_REGEXP = /^.*\.[^\\]+$/;
async function preview(config, { logging }) {
  const startServerTime = performance.now();
  const defaultOrigin = "http://localhost";
  const trailingSlash = config.devOptions.trailingSlash;
  let baseURL = new URL(config.buildOptions.site || "/", defaultOrigin);
  const staticFileServer = sirv(fileURLToPath(config.dist), {
    dev: true,
    etag: true,
    maxAge: 0
  });
  const server = http.createServer((req, res) => {
    var _a;
    const requestURL = new URL(req.url, defaultOrigin);
    if (!requestURL.pathname.startsWith(baseURL.pathname)) {
      res.statusCode = 404;
      res.end(subpathNotUsedTemplate(baseURL.pathname, requestURL.pathname));
      return;
    }
    const pathname = requestURL.pathname.slice(baseURL.pathname.length - 1);
    const isRoot = pathname === "/";
    const hasTrailingSlash = isRoot || pathname.endsWith("/");
    function sendError(message) {
      res.statusCode = 404;
      res.end(notFoundTemplate(pathname, message));
    }
    switch (true) {
      case (hasTrailingSlash && trailingSlash == "never" && !isRoot):
        sendError('Not Found (devOptions.trailingSlash is set to "never")');
        return;
      case (!hasTrailingSlash && trailingSlash == "always" && !isRoot && !HAS_FILE_EXTENSION_REGEXP.test(pathname)):
        sendError('Not Found (devOptions.trailingSlash is set to "always")');
        return;
      default: {
        req.url = "/" + ((_a = req.url) == null ? void 0 : _a.replace(baseURL.pathname, ""));
        staticFileServer(req, res, () => sendError("Not Found"));
        return;
      }
    }
  });
  let { port } = config.devOptions;
  const host = getResolvedHostForHttpServer(config);
  let httpServer;
  function startServer(timerStart) {
    let showedPortTakenMsg = false;
    let showedListenMsg = false;
    return new Promise((resolve, reject) => {
      const listen = () => {
        httpServer = server.listen(port, host, async () => {
          if (!showedListenMsg) {
            const devServerAddressInfo = server.address();
            info(logging, null, msg.devStart({ startupTime: performance.now() - timerStart, config, devServerAddressInfo, https: false, site: baseURL }));
          }
          showedListenMsg = true;
          resolve();
        });
        httpServer == null ? void 0 : httpServer.on("error", onError);
      };
      const onError = (err) => {
        if (err.code && err.code === "EADDRINUSE") {
          if (!showedPortTakenMsg) {
            info(logging, "astro", msg.portInUse({ port }));
            showedPortTakenMsg = true;
          }
          port++;
          return listen();
        } else {
          error(logging, "astro", err.stack);
          httpServer == null ? void 0 : httpServer.removeListener("error", onError);
          reject(err);
        }
      };
      listen();
    });
  }
  await startServer(startServerTime);
  return {
    host,
    port,
    server: httpServer,
    stop: async () => {
      await new Promise((resolve, reject) => {
        httpServer.close((err) => err ? reject(err) : resolve(void 0));
      });
    }
  };
}
export {
  preview as default
};
